### ISSUE TITLE
last value is always anomaly in SpectralResidual

### ISSUE THREAD

**Issue by ifokeev** [SeldonIO/alibi-detect#153]
I'm testing the library with different timeseries length and getting that last value is always an anomaly. I was playing with `n_est_points` and `n_grad_points` but no effect. In the paper https://arxiv.org/pdf/1906.03821.pdf there is something called "adjusted anomaly", but I don't see it in the lib. Is it problem on my side or it's the feature?

**Comment by arnaudvl** [comment]
Hi @ifokeev. The [SR example notebook](https://docs.seldon.io/projects/alibi-detect/en/latest/examples/od_sr_synth.html) doesn't show this behaviour, so it's a bit hard to check why it does on your data. It is possible that the extrapolation strategy used in the paper (Eq.8-9) leads to undesired behaviour. I am personally not a fan of this type of static extrapolation/padding and raised an issue to improve upon it a while ago (#71) but did not get around to it yet. The *adjusted anomaly* from the paper is just a way to evaluate the method as far as I understand (basically every anomalous segment detected *k* steps from the true anomaly is classified as fine as far as the eval metrics go).

**Comment by ifokeev** [comment]
@arnaudvl I fixed the bug just changing padding to reverse padding. Why here https://github.com/SeldonIO/alibi-detect/blob/master/alibi_detect/od/sr.py#L141 padding is concatenating plain?

**Comment by arnaudvl** [comment]
That line corresponds to Eq.9 in the paper, and the following line implements (quote from the paper) *We find that the first estimated point plays a decisive rol. Thus, we just copy x<sub>n+1</sub> for &#954; times and add the points to the tail of the sequence*. I don't however like this approach since it seems to be some heuristic that worked well for them on their datasets but has no fundamental reason and it is possible that your adjustment is way better. Would you mind sharing it? If it is a more sound extrapolation method, I would be happy to include the improvement in the library.

**Comment by ifokeev** [comment]
@arnaudvl I don't have mathematical definition about this, but I tested `SpectralResidual` with different datasets and my heuristic is working fine, default – not. Looks like the default heuristic will 100% work with a seasonal data

**Comment by arnaudvl** [comment]
Would you mind sharing the (pseudo)code so I can analyse the difference in behaviour?

**Comment by ifokeev** [comment]
```
            od = SpectralResidual(
                threshold=1.,
                window_amp=window,
                window_local=window,
                n_grad_points=10,
                n_est_points=1,
            )

            data_length = len(df.index)
            padding = -(window * 2)

            chunks_count = data_length // 2500
            if chunks_count < 2:
                chunks_count = 1

            for index_batch in np.array_split(df.index.values, chunks_count):
                new_index_batch = np.array([
                    *index_batch,
                    *np.flipud(index_batch[padding:])
                ])

                X = df.loc[new_index_batch][col].values
                t = new_index_batch.astype(np.int64) // 10**9

                od.infer_threshold(X, t, threshold_perc=100 - ANOMALIES_PERCENT)
                preds_in = od.predict(X, t, return_instance_score=return_instance_score)

```

Also I'm assuming chunks will help to identify correct mean to find outliers

**Comment by arnaudvl** [comment]
That's great, thanks. I will take a closer look and might add it as an alternative padding method.

**Comment by arnaudvl** [comment]
Addressed by #396
